feat: WalletConnect Connector

tomiir opened the pull request:
## Description

- Adds `WalletConnectWallet` standard wallet connector to connect via the WalletConnect relay
- Automatically adds the wallet if a `walletConnectProjectId` is provided in the configuration for `createDappKit`


## Test plan
- Added demo app based off enoki-connect demo 
- Manually validated supported methods

hayes-mysten commented on packages/walletconnect-wallet/src/wallet/index.ts:
```diff
@@ -0,0 +1,381 @@
+	'data:image/svg+xml;base64,PHN2ZyBmaWxsPSJub25lIiBoZWlnaHQ9IjQwMCIgdmlld0JveD0iMCAwIDQwMCA0MDAiIHdpZHRoPSI0MDAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiPjxjbGlwUGF0aCBpZD0iYSI+PHBhdGggZD0ibTAgMGg0MDB2NDAwaC00MDB6Ii8+PC9jbGlwUGF0aD48ZyBjbGlwLXBhdGg9InVybCgjYSkiPjxjaXJjbGUgY3g9IjIwMCIgY3k9IjIwMCIgZmlsbD0iIzE0MTQxNCIgcj0iMTk5LjUiIHN0cm9 ...
+
+async function getTransaction(digest: string, chain: 'mainnet' | 'testnet' | 'devnet') {
+	const client = new SuiClient({ url: getFullnodeUrl(chain) });
```
For the enoki wallet we pass in clients like this:  https://github.com/MystenLabs/ts-sdks/blob/e315e040314edc47b81bf71098e33ee941dd849d/packages/enoki/src/wallet/types.ts#L89-L102

In dapp-kit-core this can work something like this: https://github.com/MystenLabs/ts-sdks/blob/e315e040314edc47b81bf71098e33ee941dd849d/packages/enoki/src/wallet/initializer.ts#L20-L24

In the `@mysten/dapp-kit` we currently just register enoki with a single client/network that gets re-registered in network change: https://github.com/MystenLabs/ts-sdks/blob/e315e040314edc47b81bf71098e33ee941dd849d/packages/enoki/demo/RegisterEnokiWallets.tsx#L29

but this could probably also be updated to work with multiple networks

tomiir replied:
@hayes-mysten I added a `getClient` parameter. 
What I found interesting is this mechanism is already considered on `createDappKit`. Did not find the exact equivalent in react context `WalletProvider` but I did find a `SuiContext` with a `client` which I saw is updated on network switches so I made use of it. Let me know if it works 🫡  (pushing in a second)

hayes-mysten commented on packages/walletconnect-wallet/src/wallet/index.ts:
```diff
@@ -0,0 +1,381 @@
+		options: {
+			showInput: true,
+			showEffects: true,
+			showEvents: true,
```
Based on the usage below, you probably want showRawEffects and showRawInput which will give you the base64 encoded bcs bytes for the transaction and effects

hayes-mysten commented on packages/walletconnect-wallet/src/wallet/index.ts:
```diff
@@ -0,0 +1,381 @@
+
+		return {
+			digest: response.digest,
+			effects: tx.effects?.toString() ?? '',
```
tx.rawEffects is probably what you want here

hayes-mysten replied:
You should be able to use `effects: toBase64(result.effects.bcs!), ` with that api

hayes-mysten commented on packages/walletconnect-wallet/src/wallet/index.ts:
```diff
@@ -0,0 +1,381 @@
+		)) as { digest: string };
+
+		const [_, chainId] = chain.split(':');
+		const tx = await getTransaction(response.digest, chainId as 'mainnet' | 'testnet' | 'devnet');
```
rather than using this, what I would actually recommend is using the new `core` APIs.  Once your wallet has access to a `client` instance (mentioned in another comment) it should have a `core` API that will be more future proof.

The `core` methods will exist on all Sui clients (json rpc, graphql, grpc).

You should be able to use `client.core.getTransaction({ digest })` which will give you back all the needed transaction data.  The returned data has the same shape as signer.signAndExecuteTransaction and can be used similar to this: https://github.com/MystenLabs/ts-sdks/blob/e315e040314edc47b81bf71098e33ee941dd849d/packages/enoki/src/wallet/wallet.ts#L219-L230

hayes-mysten replied:
I am also realizing there is a race condition here.  You should use `waitForTransaction` instead of `getTransaction` (assuming this only returns on success) because there is a chance the RPC node the app is talking to has not picked up the new transaction after the digest is returned from the wallet

hayes-mysten commented on packages/walletconnect-wallet/src/wallet/index.ts:
```diff
@@ -0,0 +1,381 @@
+				address: account.address,
+				chains: SUI_CHAINS,
+				features: walletAccountFeatures,
+				publicKey: fromBase64(account.pubkey),
```
We should double check that this is correctly formatted.  This should be the "suiBytes" meaning the raw publicKey prefixed with the scheme bit for base keypairs.  If this also supports zklogin, passkeys, or multisig its a little more complicated

hayes-mysten commented on packages/walletconnect-wallet/CHANGELOG.md:
```diff
@@ -0,0 +1,177 @@
+# @mysten/slush-wallet
```
delete this

hayes-mysten commented on packages/walletconnect-wallet/package.json:
```diff
@@ -0,0 +1,57 @@
+		"@mysten/utils": "workspace:*",
+		"@mysten/wallet-standard": "workspace:*",
+		"@mysten/window-wallet-core": "workspace:*",
+		"@reown/appkit-universal-connector": "1.7.19-uc-config.1.0",
```
I think with prerelease external packages in this adapter, I'd prefer to avoid the direct dependency on wallet-connect in dapp-kit and dapp-kit-core at least until this has stabilized more

hayes-mysten commented on packages/walletconnect-wallet/README.md:
```diff
@@ -0,0 +1 @@
+# `@mysten/slush-wallet`
```
wrong package name

hayes-mysten commented on packages/walletconnect-wallet/package.json:
```diff
@@ -0,0 +1,57 @@
+	"files": [
+		"CHANGELOG.md",
+		"dist",
+		"experimental"
```
whats this for?

hayes-mysten commented on packages/walletconnect-wallet/typedoc.json:
```diff
@@ -0,0 +1,6 @@
+{
```
for this to be picked up in the main typedocs config, you also need an entry in the main typedoc config in the docs package

williamrobertson13 reviewed (commented):
Mostly smaller comments, I think @hayes-mysten covered all the major stuff. Still need to give it a spin locally but am very excited about this 🎉

tomiir commented on packages/dapp-kit-next/packages/dapp-kit-core/src/wallets/walletconnect.ts:
```diff
@@ -0,0 +1,17 @@
+
+export function walletConnectWalletInitializer(projectId: string): WalletInitializer {
+	return {
+		id: 'walletconnect-wallet-initializer',
```
Not a usual case but adding it in case there's some specific implementation that requires multiple instances

tomiir commented on packages/walletconnect-wallet/src/wallet/index.ts:
```diff
@@ -0,0 +1,381 @@
+			{
+				method: 'sui_signAndExecuteTransaction',
+				params: {
+					transaction: data,
```
will push to team for an update on the docs as this is already correctly parsed in our exampel wallet 🙏

tomiir commented:
@williamrobertson13 @hayes-mysten Thank you very much for your thorough reviews 🙏 

I believe I've addressed all of the issues, and for those that I haven't I left corresponding questions. 

Let me know if something is missing and I'll promptly amend! 🫡

hayes-mysten commented on packages/dapp-kit-next/examples/react/simple/src/dApp-kit.ts:
```diff
@@ -6,6 +6,7 @@ import { getFullnodeUrl, SuiClient } from '@mysten/sui/client';
 
 export const dAppKit = createDAppKit({
 	enableBurnerWallet: import.meta.env.DEV,
+	walletConnectProjectId: 'your_project_id',
```
because this requires initialization by apps anyways, I think it would be better to not have this added directly to dapp-kit, and just have apps manually add the initializer from wallet-connect-adapter instead

I would feel a lot more comfortable getting this merged if it is not automatically included in dapp-kit in its initial version.  I think we can revisit including it by default in the fututre, but for now keeping everything in the walletconnect-adapter package with an easy to use initializer should not add much work for devs, and reduces the number of things that get installed for anyone not using walletconnect

tomiir replied:
I think there is no problem with this, will remove from here and leave it up for devs to initialize. 
If there are any docs to update to list WC as an available connector, let me know!

hayes-mysten commented on packages/dapp-kit/src/components/WalletProvider.tsx:
```diff
@@ -51,6 +52,9 @@ export type WalletProviderProps = {
 	theme?: Theme | null;
 
+	/** The project id to use for the WalletConnect wallet. */
+	walletConnectProjectId?: string;
```
same comment as dapp-kit-core

hayes-mysten commented on packages/dapp-kit/src/hooks/wallet/useWalletConnectWallet.ts:
```diff
@@ -0,0 +1,36 @@
+	projectId?: string;
+}
+
+export function useWalletConnectWallet({ projectId }: WalletConnectWalletConfig) {
```
This is one of the pieces that made the dapp-kit hard to build for, and why its being replaced.  This hook really doesn't belong in dapp-kit.

Unfortunately the alternative forces you to have a peer dependency on react in the adapter package which is also not great.

There are 2 not so good options I can think of to make this work:

We could add a generic useSingleNetworkWalet that takes a register function so that someone could do something like:

```ts
import { useSingleNetworkWallet } from '@mysten/dapp-kit'
import { registerWalletConnectWallet } from '@mysten/walletconnect-wallet';
...
useSingleNetworkWalet(registerWalletConnectWallet)
```

This is probably oversimplifying things a lot, but this would enable us to keep the wallet implementations in their own packages without forcing them to deal with react directly.

The even easier (short term?) solution would be to just put this hook in the docs/setup instructions. 

I don't love it, but its one way to avoid having to solve this until we can get everything migrated to dapp-kit-core.

tomiir replied:
I think it's easier to just remove this and keep `registerWalletConnectWallet` as the main/only entrypoint and let devs instantiate as they want and not couple the dependencies.

hayes-mysten commented on packages/walletconnect-wallet/src/wallet/index.ts:
```diff
@@ -0,0 +1,406 @@
+		chain,
+	}) => {
+		const [, network] = chain.split(':');
+		const client = this.#getClient(network as Network);
```
This is potentially dangerous with the way the dapp-kit hook was implemented to just give you the current client.  I think this side is right, but if a dapp has its current network as mainnet, and then tries to sign a testnet transaction without changing the active network, the tx will be built built for the wrong network.

I think this is potentially okay, since it should still get executed against the correct network, which will cause it to error, but maybe something worth thinking about.

Once solution might be adding a check in the hook that registers the wallet that does something like:

`getClient: (network) => { if (network !== currentClient.network) { throw } return client }`

hayes-mysten reviewed (commented):
Is there a way that this can be tested?  I am not familiar with how the other side of the wallet-connect flow works

tomiir commented:
@hayes-mysten thanks again for the review 🙏 
By testing you mean as in testing coverage (unit/e2e) or as in manual testing against a wallet? 
I can deploy my changes on our example wallet if that's what you require !

hayes-mysten commented:
> By testing you mean as in testing coverage (unit/e2e) or as in manual testing against a wallet?
I can deploy my changes on our example wallet if that's what you require !

Yes, if there was an easy way to manually run through a couple flows with a demo wallet app that would be cool, not necessarily a blocker, but can be helpful to make sure everything is actually working end to end

tomiir commented:
@hayes-mysten I'm trying to get it merged but here's a preview you can use to test this 🙏 
https://react-wallet-v2-git-feat-sui-get-accounts-reown-com.vercel.app/walletconnect

hayes-mysten commented on packages/walletconnect-wallet/src/wallet/index.ts:
```diff
@@ -0,0 +1,408 @@
+	Experimental_BaseClient,
+	Experimental_SuiClientTypes,
+} from '@mysten/sui/experimental';
+import { Transaction } from '@mysten/sui/dist/cjs/transactions/Transaction';
```
```suggestion
import { Transaction } from '@mysten/transactions';
```

hayes-mysten reviewed (commented):
I think this PR looks like its just about ready to ship.  I think there are a couple of unaddressed comments, and some build failures that need to get cleaned.  Are there upstream changes we are waiting on before this can be merged and used?

tomiir commented on packages/walletconnect-wallet/src/wallet/index.ts:
```diff
@@ -0,0 +1,408 @@
+			],
+			modalConfig: {
+				themeVariables: {
+					'--w3m-z-index': 2147483647,
```
I thought I'd included this in the last version, but didn't. It will be included in next version. Using a canary image with latest published + this flag but can wait until stable is released on monday to merge this pr 🫡

tomiir commented:
Hey @hayes-mysten ! Thanks again (now for the QA 🙏 ) 

I fixed the issue with the signatures, was wrongly base64 encoding. 
Added the new version with z-index manipulation as well. Stable version including this change is planned to go live next monday, using a canary version atm so you can take a look 🫡

Wondering what other comments are still unresolved? I believe I've removed the dependency on WC from the ui level libs and left the register method alone which I think was the main point of discussion. Let me know and I can address ASAP!

hayes-mysten reviewed (approved):
Code generally looks like its in a good place once the CI checks are all passing
